system_metrics_collector/test/system_metrics_collector/test_linux_process_cpu_measurement_node.cpp

```diff
@@ -0,0 +1,222 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <memory>
+#include <fstream>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+
+#include "metrics_statistics_msgs/msg/metrics_message.hpp"
+#include "metrics_statistics_msgs/msg/statistic_data_type.hpp"
+
+#include "../../src/system_metrics_collector/linux_process_cpu_measurement_node.hpp"
+#include "../../src/system_metrics_collector/proc_cpu_data.hpp"
+#include "../../src/system_metrics_collector/utilities.hpp"
+
+#include "test_constants.hpp"
+#include "test_utilities.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+using metrics_statistics_msgs::msg::StatisticDataPoint;
+using moving_average_statistics::StatisticData;
+
+namespace
+{
+constexpr const char kTestNodeName[] = "test_measure_linux_process_cpu";
+constexpr const char kTestTopic[] = "test_process_cpu_measure_topic";
+}
+
+class TestLinuxProcessCpuMeasurementNode : public system_metrics_collector::
+  LinuxProcessCpuMeasurementNode
+{
+public:
+  TestLinuxProcessCpuMeasurementNode(
+    const std::string & name,
+    const std::chrono::milliseconds measurement_period,
+    const std::string & topic,
+    const std::chrono::milliseconds publish_period)
+  : LinuxProcessCpuMeasurementNode(name, measurement_period, topic, publish_period) {}
+
+  double PeriodicMeasurement() override
+  {
+    LinuxProcessCpuMeasurementNode::PeriodicMeasurement();
```

Comment:
Doesn't this assume that the measurement works on the machine running the integration tests? Is that an assumption we can make? I would consider stubbing out the file I/O and provide fake data.

Reply:
I think if we can't make the measurement on the machine running the integration test, then it's something wrong with the way we're measuring or something is wrong with the test machine.

Reply:
I don't agree: what happens if we need to support multiple platforms? There's no requirement that the build system is able to make these measurements, which is why many of the Linux specific functions are have been separately stubbed out and tested independently.

Reply:
But this is in `test_LINUX_process_cpu_measurement_node.cpp`

Reply:
> But this is in `test_LINUX_process_cpu_measurement_node.cpp`

Right, but the unit test makes the assumption that the machine running the test supports the linux specific measurements. I think it's important to point out none of the other Linux node tests make this assumption.

Reply:
I think having a linux specific test might make tests flakey (see buildfarm now :upside_down_face: ) or result in more configuration down the line in the CMakeLists to make sure the tests only run on specific OS/Architectures. 

If this should _only_ run on linux, then that should be specific in the CMakeLists or in the code with an `#ifdef` IMO.

This should probably fail when running on any other system otherwise.

Reply:
Per hallway discussion with @dabonnie, this should not call the linux specific component and instead use fake data used in `SetUp` or ctor function.

Reply:
> I think it's important to point out none of the other Linux node tests make this assumption.

And consequently some of the functions under `utilities.cpp` have no code coverage.

Reply:
> > I think it's important to point out none of the other Linux node tests make this assumption.
> 
> And consequently some of the functions under `utilities.cpp` have no code coverage.

We can always add tests where not making system specific calls (e.g. GetPID). I need to understand the coverage, but the majority of the utilities functions have corresponding unit tests covering their different return cases (via test_utilities.cpp)

Reply:
Oops sorry, not `utilities.cpp`, but `LinuxCpuMeasurementNode::MakeSingleMeasurement()`, `LinuxProcessMemoryMeasurementNode::PeriodicMeasurement()`, and `LinuxMemoryMeasurementNode::PeriodicMeasurement()`.

Reply:
I don't get the point you're trying to make, because we can walk through the existing unit tests and go over how all of the non-specific system calls directly tested. Otherwise one could mock these classes solely to call these methods directly, but that kind of test exists for the superclass.  

However, we're getting off topic here. Let's address the tests in this PR and improve coverage, etc in other issues.

Reply:
> how all of the non-specific system calls directly tested.

My point is why are we not testing system-specific calls?

Anyway yes, I'll disagree and commit™ to this PR comment.

Reply:
This has been mocked out via overriding `LinuxProcessCpuMeasurementNode::MakeSingleMeasurement()`.